[Feature] Implement model-free readout mitigation

rajaiitp opened the pull request:
- Worked on  [request to compute expectation values from samples #2 ](https://github.com/pasqal-io/qadence-protocols/issues/2) and  [Implement model-free readout error mitigation #5 ](https://github.com/pasqal-io/qadence-protocols/issues/5). 
- Replaced single qubit matrix multiplications without computing the complete matrix

rajaiitp committed:
faster multiplication

rajaiitp committed:
faster multiplication

rajaiitp committed:
adding measurement protocol to return expectation value from samples

rajaiitp committed:
fixing linting

RolandMacDoland commented on qadence_protocols/measurements/protocols.py:
Hey @rajaiitp everything that concerns measurements have to sit in Qadence. I tried to migrate measurement protocols here but it is not possible in state mainly for differentiability issues. We have to figure out a way but it'll take some thoughts and possibly substantial work. So everything that concerns measurements still sits in Qadence.

rajaiitp replied:
I heard we are migrating some of the QML stuff out of qadence. Maybe you can take ideas from the organization of cirq(google). Its a circuit only package with tensorflow-cirq being a QML package.

RolandMacDoland commented on qadence_protocols/mitigations/readout.py:
```diff
@@ -14,8 +13,26 @@
 
-def corrected_probas(p_corr: npt.NDArray, T: npt.NDArray, p_raw: npt.NDArray) -> np.double:
-    return norm(T @ p_corr.T - p_raw.T, ord=2) ** 2
+def tensor_rank_mult(qubit_matrices: torch.Tensor, prob_vect: npt.NDArray) -> npt.NDArray:
```
I guess a more meaningful name could help and some doctrings. I would also convert `qubit_matrices` as NDArray when called.

gvelikova replied:
@rajaiitp can you do these matrix multiplications directly with PyTorch instead?

rajaiitp replied:
@dominikandreasseitz  I'm using this tensor_rank_mult that PyQ uses in the backed i understand. Do you have a function call that i can readily use to avoid redundancy

rajaiitp replied:
Resolving the conversation for now with a comment that can be taken over if it needs to be addressed

RolandMacDoland commented on qadence_protocols/mitigations/readout.py:
```diff
@@ -14,8 +13,26 @@
+def tensor_rank_mult(qubit_matrices: torch.Tensor, prob_vect: npt.NDArray) -> npt.NDArray:
+    N = int(np.log2(len(prob_vect)))
+
+    # Reshape meausurements into a rank N tensor
```
```suggestion
    # Reshape measurements into a rank N tensor
```

RolandMacDoland commented on qadence_protocols/mitigations/readout.py:
```diff
@@ -14,8 +13,26 @@
+def corrected_probas(
+    p_corr: npt.NDArray, noise_matrices: torch.Tensor, p_raw: npt.NDArray
+) -> np.double:
+    ## computing rectified probabilites without computing the full T matrix
```
```suggestion
    # Computing rectified probabilites without computing the full T matrix
```

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -1,24 +1,30 @@
 from qadence.types import BackendName
 from torch import tensor
 
+from qadence_protocols.measurements import Measurements
```
Unfortunately, we can't do this. We have to import from Qadence.

RolandMacDoland replied:
@rajaiitp I think you have to delete all measurement related code. As I said, we can't do anything related to measurements in protocols yet.

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -1,24 +1,30 @@
     for key in samples_twirl.keys():
-        output["{0:b}".format(int(key, 2) ^ int(operand, 2))] = samples_twirl[key]
+        output[
+            ("{:0" + str(n_qubits) + "b}").format(int(key, 2) ^ int(operand, 2))
```
I think this can be recast into a syntax that avoids `.format`. Also, I would like to see it tested.

rajaiitp replied:
The expected results and inputs are not straightforward to be hard coded for the function, im not sure how to do this

rajaiitp replied:
replaced format with bin, but it also uses a certain formating "0b0111" implicity and I splice it using [2:]

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -1,24 +1,30 @@
+    return Counter(output)
 
 
 def compute_exp(n_qubits: int, samples_twirl: list, observable: list) -> tensor:
```
Docstring ? Also, needs a test.

rajaiitp replied:
removed the function, it was there as a placeholder to compute expectations. Now the samples.py protocol will do its function

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -40,7 +46,7 @@ def mitigate(
     noise: Noise,
     n_shots: int,
-    observable: list,
+    observables: list,
```
I think this has to change wrt the new API design. You can have a look at an example in `readout.py`. Basically, only the model, noise and few other arguments need to be passed, tobe able to do:

```python
mitigated = mitigate(model=model, noise=noise, ...)
```

Also, please have a look at `test_readout.py` for the usage.

RolandMacDoland commented on tests/test_twirl.py:
More tests are needed here. See above.

rajaiitp replied:
Some of these functions have been taken from qadence and modified a bit. I think its important we resolve these redundancies first

RolandMacDoland commented:
@rajaiitp Could you also provide an issue and a description of the work ?

rajaiitp committed:
adding test cases and using model as a parameter

rajaiitp commented:
Things I think require addressing

- I had to open up the model and access parameters that are not to be used i suppose. These parameters are labelled by "_xyz". 
- I also had to create a dummy circuit that is in no way used by the protocol to just get expectations from samples because of the need to create a quantum model to call the protocol
- Hardcoding circuits and observables in most cases seem to be not exhaustive and inefficient. I assume we will be changing this, right ?
- If I have missed addressing some comment, let me know. I dont have a full grasp on the qadence package and trying to fix bugs that arises from using low level function calls is becoming a pain. 

For instance, the protocol will work if  mul[3,kron(z_1,z_2)] is used in the observable, but will fail if user somehow gives mul[1,kron(3z_1,z_2)]   (which is something that pyqtorch does to simplify its computation) as the overall scaling is extracted as 1, and i dont know how to fix it when i call the function unroll pauli block with scaling coefficients. There is an evaluate function that i should be calling on the pauliterm that returns 3 independent of the structure, but I am not sure how to give it the right formatting.

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -55,10 +44,14 @@ def mitigate(
     Returns:
         Mitigated output is returned
     """
+    block = model._circuit.abstract.block
```
You might need `circuit.original` here too if you want to work with non-transpiled and non-converted circuits (this is what you did for the observables).

rajaiitp replied:
both work, but i changed it to original

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -55,10 +44,14 @@ def mitigate(
     """
+    block = model._circuit.abstract.block
 
     twirls = list(
```
Maybe a comment on what twirls are ?

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -67,23 +60,53 @@ def mitigate(
     for twirl in twirls:
         layer = [X(i) for i in twirl]
-        block_twirl = chain(circuit.block, kron(*layer))
+        block_twirl = chain(block, kron(*layer))
```
```suggestion
        block_twirl = chain(block, kron(X(i) for i in twirl))
```

Please try it but I am almost certain it works.

rajaiitp replied:
Yup it works

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -67,23 +60,53 @@ def mitigate(
-        model_twirl_num = QuantumModel(circuit=circ_twirl_num, backend=backend)
-        samples_twirl_num = model_twirl_num.sample(noise=noise, n_shots=n_shots)[0]
-        samples_twirl_num_list.append(twirl_swap(n_qubits, twirl, samples_twirl_num))
+        model_twirl_num = QuantumModel(circuit=circ_twirl_num, backend=model.backend.backend.name)
```
Is this working ? 

I thought the syntax would be something like:

```suggestion
        model_twirl_num = QuantumModel(circuit=circ_twirl_num, backend=model._backend_name)
```

rajaiitp replied:
Both work, but i changed it in accordance to your suggestion

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -67,23 +60,53 @@ def mitigate(
-        expectation_num = compute_exp(n_qubits, samples_twirl_num_list, o)
-        expectation_den = compute_exp(n_qubits, samples_twirl_den_list, o)
-        expectation += expectation_num / expectation_den
+        model_twirl_den = QuantumModel(circuit=circ_twirl_den, backend=model.backend.backend.name)
```
Ditto.

RolandMacDoland commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -67,23 +60,53 @@ def mitigate(
+
+    # I am creating a dummy circuit that has nothing to do with the protocol
+    # This needs to be changed
+    dummy = QuantumCircuit(
```
What is this for exactly ?

rajaiitp replied:
Like i mention in my comments above,  "I also had to create a dummy circuit that is in no way used by the protocol to just get expectations from samples because of the need to create a quantum model to call the protocol".

I am not sure if it needs a separate protocol. Because the protocol runs only on modified samples. Do you think I should roll back the changes and just keep it the way it was, with just adding additional function calls that helps me do the computation.

RolandMacDoland replied:
Ah, right. It feels like this is an overkill. If you only need to call expectations, I would recommend to use the convenience `expectation` function [here](https://github.com/pasqal-io/qadence/blob/acc647205e2f4b028ba7e0f699e03d384dbb1f38/qadence/execution.py#L181) in the `execution` module.  It relieves you from all the QuantumModel boilerplate.

rajaiitp replied:
This method still expect a circuit from what i see, so I would need to keep circuit still ?

rajaiitp replied:
Forgot to mark it resolved, Fixed it in the new push

RolandMacDoland reviewed (commented):
Thanks @rajaiitp getting there. Still, more tests are needed.

gvelikova committed:
remove measurements

rajaiitp committed:
fixed typos and imports

rajaiitp committed:
fixed linting

rajaiitp committed:
updating changes to circumvent model

RolandMacDoland commented:
@rajaiitp Can you have a look at resolving the conflicts ?

gvelikova committed:
Merge branch 'main' into rs/exp_values

gvelikova committed:
up the qadence version required

gvelikova commented on qadence_protocols/mitigations/readout.py:
```diff
@@ -14,8 +13,32 @@
 
-def corrected_probas(p_corr: npt.NDArray, T: npt.NDArray, p_raw: npt.NDArray) -> np.double:
-    return norm(T @ p_corr.T - p_raw.T, ord=2) ** 2
+def tensor_rank_mult(qubit_ops: torch.Tensor, prob_vect: npt.NDArray) -> npt.NDArray:
```
Here and everywhere else that you @rajaiitp use numpy instead of torch is it for faster computations, or some functionalities missing in torch?

rajaiitp replied:
@gvelikova : I did discuss this with @Roland-djee the other day, we decided to leave it as numpy unless any torch related functionalities (like gradients) are seen necessary. Let me know if you think otherwise.

gvelikova commented on qadence_protocols/mitigations/twirl.py:
```diff
@@ -1,44 +1,31 @@
 from torch import tensor
 
 
 def twirl_swap(n_qubits: int, twirl: tuple, samples_twirl: dict) -> dict:
```
I don't know if it would make sense to put more explicit typing for function arguments in general, as we did in the last qadence PR.

rajaiitp replied:
Yes, it would be good. Fixed it

RolandMacDoland commented:
> Things I think require addressing
> 
> * I had to open up the model and access parameters that are not to be used i suppose. These parameters are labelled by "_xyz".

Could you please list them ? Maybe we can make them more directly accessible in Qadence.

> * I also had to create a dummy circuit that is in no way used by the protocol to just get expectations from samples because of the need to create a quantum model to call the protocol

I think we can extend our convenience function `expectation` in Qadence to avoid this.

> * Hardcoding circuits and observables in most cases seem to be not exhaustive and inefficient. I assume we will be changing this, right ?

What exactly do you mean by this ?

> * If I have missed addressing some comment, let me know. I dont have a full grasp on the qadence package and trying to fix bugs that arises from using low level function calls is becoming a pain.
> 
> For instance, the protocol will work if mul[3,kron(z_1,z_2)] is used in the observable, but will fail if user somehow gives mul[1,kron(3z_1,z_2)] (which is something that pyqtorch does to simplify its computation) as the overall scaling is extracted as 1, and i dont know how to fix it when i call the function unroll pauli block with scaling coefficients. There is an evaluate function that i should be calling on the pauliterm that returns 3 independent of the structure, but I am not sure how to give it the right formatting.

I think we should discuss this.

rajaiitp committed:
fixing typehints and docstrings

rajaiitp commented:
Hey guys, is this good to merge ? I didnt write the test cases using random circuit generator and observables. I wasnt able to figure it out, but we can do that in a different PR maybe

RolandMacDoland commented:
> Hey guys, is this good to merge ? I didnt write the test cases using random circuit generator and observables. I wasnt able to figure it out, but we can do that in a different PR maybe

in #18.
